File: src/usecases/CreateWorkoutPlan.ts
import { NotFoundError } from "../errors/index.js";
import { WeekDay, WorkoutGoal } from "../generated/prisma/enums.js";
import { prisma } from "../lib/db.js";

// Data Transfer Object
interface InputDto {
  userId: string;
  name: string;
  goal?: WorkoutGoal;
  coverImageUrl?: string;
  workoutDays: Array<{
    name: string;
    weekDay: WeekDay;
    isRest: boolean;
    estimatedDurationInSeconds: number;
    coverImageUrl?: string;
    exercises: Array<{
      order: number;
      name: string;
      sets: number;
      reps: number;
      restTimeInSeconds: number;
    }>;
  }>;
}

interface OutputDto {
  id: string;
  name: string;
  goal?: WorkoutGoal;
  coverImageUrl?: string;
  workoutDays: Array<{
    name: string;
    weekDay: WeekDay;
    isRest: boolean;
    estimatedDurationInSeconds: number;
    coverImageUrl?: string;
    exercises: Array<{
      order: number;
      name: string;
      sets: number;
      reps: number;
      restTimeInSeconds: number;
    }>;
  }>;
}

export class CreateWorkoutPlan {
  async execute(dto: InputDto): Promise<OutputDto> {
    const coverImageUrl =
      dto.coverImageUrl ??
      dto.workoutDays.find((day) => !day.isRest && day.coverImageUrl)
        ?.coverImageUrl;
    // Transaction - Atomicidade
    return prisma.$transaction(async (tx) => {
      await tx.workoutPlan.updateMany({
        where: { userId: dto.userId, isActive: true },
        data: { isActive: false },
      });
      const workoutPlan = await tx.workoutPlan.create({
        data: {
          id: crypto.randomUUID(),
          name: dto.name,
          userId: dto.userId,
          goal: dto.goal,
          coverImageUrl,
          isActive: true,
          workoutDays: {
            create: dto.workoutDays.map((workoutDay) => ({
              name: workoutDay.name,
              weekDay: workoutDay.weekDay,
              isRest: workoutDay.isRest,
              estimatedDurationInSeconds: workoutDay.estimatedDurationInSeconds,
              coverImageUrl: workoutDay.coverImageUrl,
              exercises: {
                create: workoutDay.exercises.map((exercise) => ({
                  name: exercise.name,
                  order: exercise.order,
                  sets: exercise.sets,
                  reps: exercise.reps,
                  restTimeInSeconds: exercise.restTimeInSeconds,
                })),
              },
            })),
          },
        },
      });
      const result = await tx.workoutPlan.findUnique({
        where: { id: workoutPlan.id },
        include: {
          workoutDays: {
            include: {
              exercises: true,
            },
          },
        },
      });
      if (!result) {
        throw new NotFoundError("Workout plan not found");
      }
      return {
        id: result.id,
        name: result.name,
        goal: result.goal ?? undefined,
        coverImageUrl: result.coverImageUrl ?? undefined,
        workoutDays: result.workoutDays.map((day) => ({
          name: day.name,
          weekDay: day.weekDay,
          isRest: day.isRest,
          estimatedDurationInSeconds: day.estimatedDurationInSeconds,
          coverImageUrl: day.coverImageUrl ?? undefined,
          exercises: day.exercises.map((exercise) => ({
            order: exercise.order,
            name: exercise.name,
            sets: exercise.sets,
            reps: exercise.reps,
            restTimeInSeconds: exercise.restTimeInSeconds,
          })),
        })),
      };
    });
  }
}

File: backend/FitAi.Api/UseCases/WorkoutPlans/CreateWorkoutPlan.cs
using FitAi.Api.Data;
using FitAi.Api.Entities;
using FitAi.Api.Errors;
using FitAi.Api.UseCases.Shared;
using FitAi.Contracts;
using Microsoft.EntityFrameworkCore;

namespace FitAi.Api.UseCases.WorkoutPlans;

/// <summary>
/// Cria um plano e o torna o plano ativo do aluno, desativando o anterior.
/// Usado pelo Coach AI (origem AI) e pelo professor no painel (origem TEACHER).
/// </summary>
public sealed class CreateWorkoutPlan(AppDbContext db)
{
    public sealed record Input(string UserId, SaveWorkoutPlanRequest Plan, WorkoutPlanSource Source, string CreatedById);

    public async Task<WorkoutPlanResponse> ExecuteAsync(Input input, CancellationToken ct = default)
    {
        WorkoutPlanValidator.Validate(input.Plan);

        var userExists = await db.Users.AnyAsync(u => u.Id == input.UserId, ct);
        if (!userExists) throw new NotFoundException("User not found");

        // Um único SaveChanges: desativar o plano atual e criar o novo acontecem na mesma transação.
        var activePlans = await db.WorkoutPlans.Where(p => p.UserId == input.UserId && p.IsActive).ToListAsync(ct);
        foreach (var activePlan in activePlans) activePlan.IsActive = false;

        var days = input.Plan.WorkoutDays.ToEntities().ToList();
        var plan = new WorkoutPlan
        {
            Name = input.Plan.Name.Trim(),
            UserId = input.UserId,
            Goal = input.Plan.Goal,
            CoverImageUrl = string.IsNullOrWhiteSpace(input.Plan.CoverImageUrl)
                ? days.FirstOrDefault(d => !d.IsRest && d.CoverImageUrl is not null)?.CoverImageUrl
                : input.Plan.CoverImageUrl,
            IsActive = true,
            Source = input.Source,
            CreatedById = input.CreatedById,
            WorkoutDays = days,
        };
        db.WorkoutPlans.Add(plan);
        await db.SaveChangesAsync(ct);

        return plan.ToResponse();
    }
}

public static class WorkoutPlanValidator
{
    public static void Validate(SaveWorkoutPlanRequest plan)
    {
        if (string.IsNullOrWhiteSpace(plan.Name)) throw new ValidationException("Informe o nome do plano");
        if (plan.WorkoutDays.Count == 0) throw new ValidationException("O plano precisa ter pelo menos um dia");

        var duplicated = plan.WorkoutDays.GroupBy(d => d.WeekDay).FirstOrDefault(g => g.Count() > 1);
        if (duplicated is not null) throw new ValidationException($"O dia {duplicated.Key} aparece mais de uma vez");

        foreach (var day in plan.WorkoutDays)
        {
            if (string.IsNullOrWhiteSpace(day.Name)) throw new ValidationException("Informe o nome de cada dia");
            if (!day.IsRest && day.Exercises.Count == 0)
            {
                throw new ValidationException($"O treino '{day.Name}' precisa de pelo menos um exercício");
            }
            if (!day.IsRest && day.EstimatedDurationInSeconds <= 0)
            {
                throw new ValidationException($"Informe a duração do treino '{day.Name}'");
            }
            if (day.Exercises.Any(e => string.IsNullOrWhiteSpace(e.Name) || e.Sets < 1 || e.Reps < 1 || e.RestTimeInSeconds < 0))
            {
                throw new ValidationException($"O treino '{day.Name}' tem um exercício inválido (séries e repetições devem ser maiores que zero)");
            }
        }
    }
}
